### DISCUSSION TITLE
S0 and Non-Secure devices reporting Security 2 CC support

### DISCUSSION THREAD

**Discussion by kpine** [zwave-js/zwave-js#3465] [Q&A] [answered]
I have a couple of devices (at least) that do not support Security 2, yet zwave-js is reporting that they do.

For example, in the HA dump I see this:
```text
            "commandClasses": [
<snip>
              {
                "id": 159,
                "name": "Security 2",
                "version": 1,
                "isSecure": true
              },
```
This dump is for a GE/Jasco 14287 / ZW4002 In-Wall Smart Fan Control, which doesn't support security at all. I have also seen the CC reports with `"isSecure"` set to false, re-interviewing seems to change it. The device does not report any Security CC in the NIF:

```text
2021-10-02T17:27:53.953Z DRIVER « [Node 024] [REQ] [ApplicationUpdateRequest]
                                    payload: 0x8418110411085e86725a855973262770562c2b7a
2021-10-02T17:27:53.957Z CNTRLR « [Node 024] node info received
                                  supported CCs:
                                  · Z-Wave Plus Info
                                  · Version
                                  · Manufacturer Specific
                                  · Device Reset Locally
                                  · Association
                                  · Association Group Information
                                  · Powerlevel
                                  · Multilevel Switch
                                  · All Switch
                                  · Configuration
                                  · CRC-16 Encapsulation
                                  · Scene Actuator Configuration
                                  · Scene Activation
                                  · Firmware Update Meta Data
                                  controlled CCs:
2021-10-02T17:27:53.962Z CNTRLR   [Node 024] Interview stage completed: NodeInfo
```
Yet later on, it seems to think it's supported:

```text
2021-10-02T17:27:54.367Z CNTRLR   [Node 024]   supports CC Manufacturer Specific (0x72) in version 2
2021-10-02T17:27:54.368Z CNTRLR   [Node 024]   skipping query for Powerlevel (0x73) because max implemented vers
                                  ion is 0
2021-10-02T17:27:54.368Z CNTRLR   [Node 024]   skipping query for Security 2 (0x9f) because max implemented vers
                                  ion is 1
```

Is this a device issue, or driver? Nothing to worry about? I am assuming the HA dump is correct, because the re-interview logs show it mentioning Security S2 CC version for those nodes (I didn't check the cache). I have another device exhibiting the same behavior, the Aeotec Gen5 Garage Door opener. That supports S0, but not S2. Another GE dimmer switch does not have this problem.

Node 022 - Aeotec Gen5 GDO, S0 but no S2 --> zwave-js thinks S2 and S0 are supported
Node 024 - GE Fan Controller, no Security supported --> zwave-js thinks S2 (not S0) are supported
Node 002 - GE Dimmer, no Security supported --> zwave-js doesn't think any Security CC are supported

[interviews.log](https://github.com/zwave-js/node-zwave-js/files/7271946/interviews.log)
[zwave_js_dump.json.txt](https://github.com/zwave-js/node-zwave-js/files/7271948/zwave_js_dump.json.txt)
[cache.zip](https://github.com/zwave-js/node-zwave-js/files/7271967/cache.zip)

**Comment by kpine** [comment, marked as answer]
OK, to possibly answer my own question, I'm assuming this has to do with the device classes. The devices.json file says that Fan Controllers and Barrier Devices Z-Wave device types (and others) implicitly support S2. For some reason multilevel switches don't have this requirement though? That would explain why it doesn't show up for the dimmer.

The spec does say:
> All Z-Wave nodes certified after the 2nd of April 2017 MUST support the Security 2 Command Class.

And S2 is listed as a mandatory CC. As we all know, devices often don't follow the spec, and they seem to have been certified anyways.

Given that there's no Security 2 CC advertisement in the NIF, does adding the CC have any effect on how the device is included (I have not tried)? Would there be any purpose in removing this CCs from the device via the config files, or is it benign, and I can just ignore?

**Reply by AlCalzone** [reply]
Your assessment is correct. Unfortunately when it comes to inclusion, we cannot identify the device before the security bootstrapping, so config files are useless. All we have to go by are the NIF and the mandatory CCs of the device class. The requirements for the latter likely changed over time, but the standard doesn't give us any way to distinguish older from new devices at that stage.
The later queries during the interview are influenced by the config files, so it probably won't hurt to remove the CC there.

**Reply by kpine** [reply]
Does that mean that for my Z-Wave+ Garage Door Opener, which does not support S2, the driver would attempt to bootstrap it with S2 (in default strategy) because Security 2 is listed as a mandatory CC for the device type? Even if not reported in the NIF?

**Reply by AlCalzone** [reply]
Yes

**Reply by kpine** [reply]
Ok, so I did confirm this makes my GDO include as insecure, which isn't a great "out of the box" experience (_it's old though, who cares I guess?_). This requirement seems to be inconsistently enforced by the driver (according to the `deviceClasses.json` file). None of the other "requires security" devices, like Locks, Keypads, etc, are forced to have Security 2 as a supported CC. Why are the Barrier devices handled differently? Can it be removed from all the "requires security" devices to avoid this problem?

It's also inconsistently enforced for non-security devices. Multilevel Switch, Binary Switch, Thermostats, Windows Covers, etc. aren't forced to support S2, but a Color Bulb does? It's not such a big deal for these devices that can function w/o security, but does it produce an error about low security?

**Reply by AlCalzone** [reply]
> It's also inconsistently enforced for non-security devices

Welcome to the Z-Wave standard 🙄

I agree, it makes sense to drop Security S0/S2 from the mandatory CCs for the reasons you mentioned.

**Reply by AlCalzone** [reply]
merged into master
